artiq/compiler/transforms/artiq_ir_generator.py

```diff
@@ -1198,7 +1198,28 @@ def visit_SubscriptT(self, node):
         finally:
             self.current_assign = old_assign
 
-        if isinstance(node.slice, ast.Index):
+        if types.is_tuple(node.value.type):
+            assert isinstance(node.slice, ast.Index), \
+                "Internal compiler error: tuple index should be an Index"
+            assert isinstance(node.slice.value, ast.Num), \
+                "Internal compiler error: tuple index should be a constant"
+
+            if self.current_assign is not None:
+                diag = diagnostic.Diagnostic("error",
+                    "cannot assign to a tuple element",
+                    {}, node.loc)
+                self.engine.process(diag)
+                return
```

Comment:
return after ``engine.process()`` is not necessary - the diagnostic engine raises an error anyway